src/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFile.cs

```diff
@@ -550,21 +558,34 @@ private static string EntryFromPath(string entry, int offset, int length)
             }
 
             if (length == 0)
-                return String.Empty;
+                return appendPathSeparator ? PathSeparatorString : String.Empty;
 
-            // create a mutable copy
-            char[] chars = entry.ToCharArray(offset, length);
+            // Create a new resulting string instance of the appropriate substring length. The string is
+            // initially populated with PathSeparator ('/'), but will be filled in (mutated) with the actual
+            // content of the substring below, before it is returned. Path.DirectorySeparatorChar and
+            // AltDirectorySeparatorChar are skipped so that PathSeparator remains in their place, in
+            // accordance with the .ZIP File Format Specification 4.4.17.1, which states all slashes should
+            // be forward slashes. The last character remains PathSeparator if appendPathSeparator is true.
+            String result = new String(PathSeparator, appendPathSeparator ? length + 1 : length);
 
-            // '/' is a more broadly recognized directory separator on all platforms (eg: mac, linux)
-            // We don't use Path.DirectorySeparatorChar or AltDirectorySeparatorChar because this is
-            // explicitly trying to standardize to '/'
-            for(int i = 0; i < chars.Length; i++)
+            unsafe
             {
-                if (chars[i] == Path.DirectorySeparatorChar || chars[i] == Path.AltDirectorySeparatorChar)
-                    chars[i] = PathSeparator;
+                fixed (char* pResult = result)
```

Comment:
Thanks for the feedback, guys. In lieu of having "something along the lines of #180 for string creation" readily available, I'll update this PR with the alternate approach I mentioned, which doesn't use unsafe code.

> I believe you'd get a speedup by passing '\0' to the constructor

Now that you mention it, this change to string sounds familiar. Thanks for the reminder!